File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//     pub val: i32,
//     pub next: Option<Box<ListNode>>,
// }
//
// impl ListNode {
//     #[inline]
//     pub fn new(val: i32) -> Self {
//         ListNode { next: None, val }
//     }
// }

impl Solution {
    pub fn reverse_list(mut head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut prev = None;

        while let Some(mut cur) = head {
            head = cur.next.take();

            cur.next = prev;

            prev = Some(cur);
        }
        prev
    }
}

File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode res;
        ListNode* res_cur = &res;

        while (list1 != nullptr && list2 != nullptr) {
            if (list1->val <= list2->val) {
                res_cur->next = list1;
                list1 = list1->next;
            } else {
                res_cur->next = list2;
                list2 = list2->next;
            }

            res_cur = res_cur->next;
        }

        if (list1 == nullptr) {
            res_cur->next = list2;
        } else {
            res_cur->next = list1;
        }

        return res.next;
    }
};
